cmd/car/compile.go

```diff
@@ -0,0 +1,446 @@
+	ls := cidlink.DefaultLinkSystem()
+	store := memstore.Store{Bag: map[string][]byte{}}
+	ls.SetReadStorage(&store)
+	ls.SetWriteStorage(&store)
+	b := basicnode.Prototype.Any.NewBuilder()
+	if encoding == "json" {
+		if err := dagjson.Decode(b, bytes.NewBuffer(raw)); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else if encoding == "raw" {
+		if err := b.AssignBytes(raw); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else {
+		return cid.Undef, nil, fmt.Errorf("unknown encoding: %s", encoding)
+	}
+	lnk, err := ls.Store(linking.LinkContext{Ctx: ctx}, cidlink.LinkPrototype{Prefix: codec}, b.Build())
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+	outCid := lnk.(cidlink.Link).Cid
+	outBytes, outErr := store.Get(ctx, outCid.KeyString())
+	return outCid, outBytes, outErr
+}
+
+// DebugCar is a command to translate between a car file, and a human-debuggable patch-like format.
+func DebugCar(c *cli.Context) error {
+	var err error
+	inStream := os.Stdin
+	inFile := "-"
+	if c.Args().Len() >= 1 {
+		inFile = c.Args().First()
+		inStream, err = os.Open(inFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	rd, err := carv2.NewBlockReader(inStream)
+	if err != nil {
+		return err
+	}
+
+	// patch the header.
+	outStream := os.Stdout
+	if c.IsSet("output") {
+		outFileName := c.String("output")
+		outFile, err := os.Create(outFileName)
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+		outStream = outFile
+	}
+
+	outStream.WriteString("car compile ")
+	if rd.Version == 2 {
+		outStream.WriteString("--v2 ")
+	}
+
+	outStream.WriteString(inFile + "\n")
+	for _, rt := range rd.Roots {
+		fmt.Fprintf(outStream, "root %s\n", rt.String())
+	}
+
+	// patch each block.
+	nxt, err := rd.Next()
+	if err != nil {
+		return err
+	}
+	for nxt != nil {
+		chunk, err := patch(c.Context, nxt.Cid(), nxt.RawData())
+		if err != nil {
+			return err
+		}
+		outStream.Write(chunk)
+
+		nxt, err = rd.Next()
+		if err == io.EOF {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func patch(ctx context.Context, c cid.Cid, blk []byte) ([]byte, error) {
+	ls := cidlink.DefaultLinkSystem()
+	store := memstore.Store{Bag: map[string][]byte{}}
+	ls.SetReadStorage(&store)
+	ls.SetWriteStorage(&store)
+	store.Put(ctx, c.KeyString(), blk)
+	node, err := ls.Load(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: c}, basicnode.Prototype.Any)
+	if err != nil {
+		return nil, fmt.Errorf("could not load block: %q", err)
+	}
+
+	outMode := "json"
+	if node.Kind() == datamodel.Kind_Bytes && isPrintable(node) {
+		outMode = "raw"
+	}
+	finalBuf := bytes.NewBuffer(nil)
+
+	if outMode == "json" {
+
+		initalJson := bytes.NewBuffer(nil)
+		if err := dagjson.Encode(node, initalJson); err != nil {
+			return nil, err
+		}
+		// re-do it with standard json to pretty print it.
+		if err := json.Indent(finalBuf, initalJson.Bytes(), "", "  "); err != nil {
+			return nil, err
+		}
```

Comment:
so .. this is a bit naughty but you can pull this off in one pass with go-ipld-prime by calling `Marshal` and passing in some refmt options, see https://go.dev/play/p/-OCdX_w5l52

Reply:
actually it's slightly more complicated to make proper dag-json, I forgot the options to encode links, bytes and sort maps: https://go.dev/play/p/XyfcQAzhzpw